Add basic submission dashboard

msmolens opened the pull request:
This adds an administrator interface to view all submissions to a phase. A
button at the bottom of the phase page routes to this interface.

Currently, the view shows submissions grouped by participant. By default,
submissions for up to five participants are shown at a time. The controls at the
top of the interface page through the list. The 'participantLimit' query
parameter sets the number of participants per page.

By default, up to five submissions per participant are shown at a time. The
controls in the group for each user page through the list. The 'submissionLimit'
query parameter sets the number of submissions to show per page.

Each submission shows its title, whether it's the latest successful submission,
its job status, its creation date, and its overall score. The job status text
links to the job detail page. Unsuccessful jobs include a tooltip showing the
first few lines of the job log. The overall score text links to the score detail
page. Links are provided to download the submission data and to view the raw
submission JSON.

To provide some extra information, warning messages are shown when a user hasn't
submitted at all or when the current page shows no successful submissions.

Some useful extensions, perhaps requiring backend enhancements to be efficient,
to this could include:
- display success/error/total submission counts per user/phase/challenge
- identify and group participants who have no successful submissions
- identify and group participants who haven't submitted
- filtering/sorting tables

Fixes #186 

![all_submissions](https://cloud.githubusercontent.com/assets/7122091/15058255/35a7eb62-12e9-11e6-94d8-6d44cc64b484.png)

zachmullen commented on web_external/js/views/body/PhaseSubmissionsView.js:
```diff
@@ -0,0 +1,86 @@
+covalic.router.route('phase/:id/submissions', 'phaseSubmissions', function (id, params) {
+    var phase = new covalic.models.PhaseModel({
+        _id: id
+    }).on('g:fetched', function () {
```
Let's use `.once` instead of `.on` so the binding doesn't stick around.

zachmullen commented on web_external/js/views/body/UserSubmissionsView.js:
```diff
@@ -0,0 +1,90 @@
+                    jobs.push(job);
+                    promises.push(deferred.promise());
+                });
+                $.when.apply($, promises).done(_.bind(function () {
```
Isn't `$.when.apply($, promises)` equivalent to `$.when(promises)`?

zachmullen replied:
Nevermind...

> If a single argument is passed to jQuery.when() and it is not a Deferred or a Promise, it will be treated as a resolved Deferred and any doneCallbacks attached will be executed immediately.

zachmullen commented on web_external/templates/body/userSubmissionsPage.jade:
```diff
@@ -0,0 +1,90 @@
+-     shortLog.push('...')
+-   }
+-   return shortLog.join('\n')
+-  }
```
I vote we move this function definition out of the jade template and pass it into the template as a variable.

msmolens commented:
@zachmullen Thanks for the review! PTAL at the changes.

zachmullen commented:
I added one commit that just cleans up some of our template logic, otherwise this looks great and works. If my code changes look good to you, merge away.

zachmullen commented:
Actually I forgot, there was one bug I noticed, which is that the "view raw" doesn't work due to not authenticating the user. We could add the cookie support to the endpoint, but it might be easier to just make the XHR ourselves if they click that link and display it in a popover or something.

msmolens commented:
I see, the endpoint that "view raw" used worked for public phases, but not private phases. I added a commit that instead dumps the JSON to the console. Since this is an administrator interface this seems sufficient and useful.

I also fixed a problem when logged in as a user with view access, but not admin access, to the phase. In this case the job resources (and therefore job statuses) aren't accessible. Now, the jobs aren't fetched or linked, and non-successful statuses just appear blank.

zachmullen commented:
LGTM
